lib/spack/spack/container/writers/__init__.py

```diff
@@ -132,10 +132,41 @@ def os_packages(self):
 
     @tengine.context_property
     def extra_instructions(self):
-        Extras = collections.namedtuple('Extra', ['build', 'final'])
+        Extras = collections.namedtuple('Extra', ['setup', 'build', 'final'])
         extras = self.container_config.get('extra_instructions', {})
+        setup = extras.get('setup', None)
         build, final = extras.get('build', None), extras.get('final', None)
-        return Extras(build=build, final=final)
+        return Extras(setup=setup, build=build, final=final)
+
+    @tengine.context_property
+    def from_host_to_build(self):
+        """Files that needs to be copied from the host to the build stage."""
+        return self._files_to_be_copied('build')
+
+    @tengine.context_property
+    def from_host_to_final(self):
+        """Files that needs to be copied from the host to the build stage."""
+        files = self._files_to_be_copied('final')
+        files = [x for x in files if not x.source.startswith('build:')]
+        return files
+
+    @tengine.context_property
+    def from_build_to_final(self):
+        """Files that needs to be copied from the host to the build stage."""
+        files = self._files_to_be_copied('final')
+        files = [x for x in files if x.source.startswith('build:')]
```

Comment:
Maybe strip the `build:` in the code?

```suggestion
        files = [x[6:] for x in files if x.source.startswith('build:')]
```